third_party/move/move-compiler-v2/src/env_pipeline/lambda_lifter.rs

```diff
@@ -0,0 +1,322 @@
+        self.fun_env
+            .module_env
+            .env
+            .symbol_pool()
+            .make(&format!("param${}", parameter_pos))
+    }
+
+    fn gen_closure_function_name(&mut self) -> Symbol {
+        let env = self.fun_env.module_env.env;
+        env.symbol_pool().make(&format!(
+            "{}$lambda${}",
+            self.fun_env.get_name().display(env.symbol_pool()),
+            self.lifted.len() + 1
+        ))
+    }
+
+    fn bind(&self, mut bindings: Vec<(Pattern, Exp)>, exp: Exp) -> Exp {
+        if let Some((pat, binding)) = bindings.pop() {
+            let env = self.fun_env.module_env.env;
+            let body = self.bind(bindings, exp);
+            let loc = env.get_node_loc(pat.node_id());
+            let ty = env.get_node_type(body.node_id());
+            let new_id = env.new_node(loc, ty);
+            ExpData::Block(new_id, pat, Some(binding), body).into_exp()
+        } else {
+            exp
+        }
+    }
+}
+
+impl<'a> ExpRewriterFunctions for LambdaLifter<'a> {
+    fn rewrite_exp(&mut self, exp: Exp) -> Exp {
+        // Intercept descent and compute lambdas being exempted from lifting, currently
+        // those passed as parameters to inline functions.
+        if !self.options.include_inline_functions {
+            if let ExpData::Call(_, Operation::MoveFunction(mid, fid), args) = exp.as_ref() {
+                let env = self.fun_env.module_env.env;
+                if env.get_function(mid.qualified(*fid)).is_inline() {
+                    for arg in args {
+                        if matches!(arg.as_ref(), ExpData::Lambda(..)) {
+                            self.exempted_lambdas.insert(arg.node_id());
+                        }
+                    }
+                }
+            }
+        }
+        // Also if this is a lambda, before descent, clear any usages from siblings in the
+        // context, so we get the isolated usage information for the lambda's body.
+        if matches!(exp.as_ref(), ExpData::Lambda(..)) {
+            let mut curr_free_params = mem::take(&mut self.free_params);
+            let mut curr_free_locals = mem::take(&mut self.free_locals);
+            let result = self.rewrite_exp_descent(exp);
+            self.free_params.append(&mut curr_free_params);
+            self.free_locals.append(&mut curr_free_locals);
+            result
+        } else {
+            self.rewrite_exp_descent(exp)
+        }
+    }
+
+    fn rewrite_enter_scope<'b>(&mut self, vars: impl Iterator<Item = &'b (NodeId, Symbol)>) {
+        self.scopes
+            .push(vars.map(|(_, sym)| sym).cloned().collect())
+    }
+
+    fn rewrite_exit_scope(&mut self) {
+        let exiting = self.scopes.pop().expect("stack balanced");
+        // Remove all locals which are bound in the scope we are exiting.
+        self.free_locals.retain(|name, _| !exiting.contains(name));
+    }
+
+    fn rewrite_local_var(&mut self, id: NodeId, sym: Symbol) -> Option<Exp> {
```

Comment:
You also need to implement `rewrite_pattern` and catch the case `matches!(pat, Pattern::Var(id, sym))` and `entering_scope=false` and record it in `self.free_modified_syms`.  You will have to subtract exiting syms from it in `rewrite_exit_scope` above, and then do a second rewrite pass to turn all free modified syms into `&mut` params and modify all uses to read/write through the references.  Note that both `Temporary` and `Parameter` can show up as a `Pattern::Var`, but that's a tiny part of the problem.

Reply:
Good catch, but I'm not going for this complexity, at least for now. When a variable is assigned to our mutably borrowed, I produce now an error.

This is similar as in Java and C# lambdas, and I think C++ has this too. We may go back and make this more powerful, but lets start simple here. Indeed, for inlined lambdas this restriction doesn't exist, but they also have other stuff which can be done which can't be in a regular lambda.

Reply:
Ok, then you need an error in this case.  Ok, see it.